pkg/virt-controller/ipamclaims/ipamclaims.go

```diff
@@ -0,0 +1,208 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2024 Red Hat, Inc.
+ *
+ */
+
+package ipamclaims
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	virtv1 "kubevirt.io/api/core/v1"
+
+	"kubevirt.io/kubevirt/pkg/network/namescheme"
+	"kubevirt.io/kubevirt/pkg/network/vmispec"
+	ipct "kubevirt.io/kubevirt/pkg/virt-controller/ipamclaims/types"
+	"kubevirt.io/kubevirt/pkg/virt-controller/network"
+
+	networkv1 "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/apis/k8s.cni.cncf.io/v1"
+
+	ipamclaims "github.com/k8snetworkplumbingwg/ipamclaims/pkg/crd/ipamclaims/v1alpha1"
+	ipamclaimsclient "github.com/k8snetworkplumbingwg/ipamclaims/pkg/crd/ipamclaims/v1alpha1/apis/clientset/versioned"
+
+	networkclient "kubevirt.io/client-go/generated/network-attachment-definition-client/clientset/versioned"
+)
+
+type IPAMClaimsManager struct {
+	networkClient    networkclient.Interface
+	ipamClaimsClient ipamclaimsclient.Interface
+}
+
+func NewIPAMClaimsManager(networkClient networkclient.Interface, ipamClaimsClient ipamclaimsclient.Interface) *IPAMClaimsManager {
+	return &IPAMClaimsManager{
+		networkClient:    networkClient,
+		ipamClaimsClient: ipamClaimsClient,
+	}
+}
+
+func (m *IPAMClaimsManager) CreateIPAMClaims(namespace string, vmiName string, interfaces []virtv1.Interface, networks []virtv1.Network, ownerRef *v1.OwnerReference) error {
+	nonAbsentNetworks := filterNonAbsentNetworks(interfaces, networks)
+	networkToIPAMClaimParams, err := m.GetNetworkToIPAMClaimParams(
+		namespace,
+		vmiName,
+		vmispec.FilterMultusNonDefaultNetworks(nonAbsentNetworks))
+	if err != nil {
+		return fmt.Errorf("failed composing networkToIPAMClaimName: %w", err)
+	}
+
+	claims := composeIPAMClaims(namespace, ownerRef, networkToIPAMClaimParams)
+	if err := m.createIPAMClaims(namespace, claims); err != nil {
+		return fmt.Errorf("failed IPAMClaims creation for VMI %s: %w", vmiName, err)
+	}
+
+	return nil
+}
+
+func composeIPAMClaims(namespace string, ownerRef *v1.OwnerReference, networkToIPAMClaimParams map[string]ipct.IPAMClaimParams) []*ipamclaims.IPAMClaim {
+	claims := []*ipamclaims.IPAMClaim{}
+	for netName, ipamClaimParams := range networkToIPAMClaimParams {
+		claims = append(claims, composeIPAMClaim(
+			namespace,
+			*ownerRef,
+			ipamClaimParams,
+			namescheme.GenerateHashedInterfaceName(netName),
```

Comment:
@oshoval did you act on this ? 

@AlonaKaplan is there a label / annotation we can use ? Or we need to check the interface name, and identify which naming scheme was used to create it ?

Reply:
Yes, this is why i changed the function to CreateNewPodIPAMClaims to reflect it is limited to new pods only
(therefore it can use the new naming scheme)

wrt to annotations, the blocking was done on VM level